File: src/test/java/ch/uzh/ifi/access/service/CourseServiceTests.java
package ch.uzh.ifi.access.service;

import ch.uzh.ifi.access.model.Assignment;
import ch.uzh.ifi.access.model.Course;
import ch.uzh.ifi.access.model.Task;
import ch.uzh.ifi.access.model.TaskFile;
import ch.uzh.ifi.access.repository.AssignmentRepository;
import ch.uzh.ifi.access.repository.TaskFileRepository;
import ch.uzh.ifi.access.repository.TaskRepository;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;

import javax.transaction.Transactional;
import java.time.LocalDate;
import java.time.LocalDateTime;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

@SpringBootTest
@Transactional
class CourseServiceTests {

    @Autowired
    private CourseService courseService;

    @Autowired
    private AssignmentRepository assignmentRepository;

    @Autowired
    private TaskRepository taskRepository;

    @Autowired
    private TaskFileRepository taskFileRepository;

    @Test
    void createCourseFromRepositoryTest() {
        String repository = "https://github.com/mp-access/Mock-Course.git";
        Course createdCourse = courseService.createCourseFromRepository(repository);
        assertEquals("Mock Course", createdCourse.getTitle());
        assertEquals(LocalDate.of(2020, 1, 1), createdCourse.getEndDate());
        assertEquals(3, createdCourse.getAssignments().size());
        Assignment createdAssignment = createdCourse.getAssignments().stream().filter(assignment ->
                assignment.getOrdinalNum().equals(1)).findFirst().orElseThrow();
        assertEquals(6, createdAssignment.getTasks().size());
        assertEquals("Introduction to Python", createdAssignment.getTitle());
        assertEquals(LocalDateTime.of(2050, 12, 4, 8, 0), createdAssignment.getEndDate());
        Task codeTask = createdAssignment.getTasks().stream().filter(task ->
                task.getOrdinalNum().equals(1)).findFirst().orElseThrow();
        assertTrue(codeTask.getInstructions().startsWith("In this task you will model the information system"));
        assertEquals("UZH Airlines", codeTask.getTitle());
        assertEquals(10, codeTask.getMaxPoints());
        assertEquals(8, codeTask.getFiles().size());
    }

    @Test
    void updateCourseFromRepositoryTest() {
        String repository = "https://github.com/mp-access/Mock-Course.git";
        Course createdCourse = courseService.createCourseFromRepository(repository);
        Assignment createdAssignment = assignmentRepository.findByCourse_UrlAndOrdinalNum(
                createdCourse.getUrl(), 1).orElseThrow();
        Task codeTask = taskRepository.findByAssignment_IdAndOrdinalNum(createdAssignment.getId(), 1).orElseThrow();
        String originalDescription = codeTask.getInstructions();
        codeTask.setInstructions("Changed description");
        taskRepository.save(codeTask);
        TaskFile gradingFile = taskFileRepository.findByTask_IdAndPath(codeTask.getId(), "private/testSuite.py").orElseThrow();
        String originalTemplate = gradingFile.getTemplate();
        gradingFile.setTemplate("Changed template");
        taskFileRepository.save(gradingFile);
        Course updatedCourse = courseService.updateCourseFromRepository(createdCourse.getUrl());
        Assignment updatedAssignment = updatedCourse.getAssignments().stream()
                .filter(assignment -> assignment.getOrdinalNum().equals(1)).findFirst().orElseThrow();
        assertEquals(createdAssignment.getId(), updatedAssignment.getId());
        Task updatedCodeTask = taskRepository.findByAssignment_IdAndOrdinalNum(updatedAssignment.getId(), 1).orElseThrow();
        assertEquals(codeTask.getId(), updatedCodeTask.getId());
        assertEquals(originalDescription, updatedCodeTask.getInstructions());
        TaskFile updatedGradingFile = taskFileRepository.findByTask_IdAndPath(
                updatedCodeTask.getId(), "private/testSuite.py").orElseThrow();
        assertEquals(gradingFile.getId(), updatedGradingFile.getId());
        assertEquals(originalTemplate, updatedGradingFile.getTemplate());
    }

}

File: src/test/kotlin/ch/uzh/ifi/access/service/CourseServiceTests.kt
package ch.uzh.ifi.access.service

import ch.uzh.ifi.access.model.Course
import ch.uzh.ifi.access.projections.CourseWorkspace
import com.fasterxml.jackson.databind.json.JsonMapper
import org.hibernate.Hibernate
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.MethodOrderer
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestMethodOrder
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.security.test.context.support.WithMockUser
import org.springframework.test.web.servlet.MockMvc
import org.springframework.web.server.ResponseStatusException
import java.io.File
import java.nio.file.Paths
import java.time.LocalDateTime


@AutoConfigureMockMvc
@SpringBootTest
@TestMethodOrder( MethodOrderer.MethodName::class)
class CourseLifecycleTests(
    @Autowired val mvc: MockMvc,
    @Autowired val jsonMapper: JsonMapper,
    @Autowired val courseLifecycle: CourseLifecycle,
    @Autowired val courseService: CourseService) {

    @Test
    @WithMockUser(username="supervisor@uzh.ch", authorities = ["supervisor"])
    fun `0_Course import succeeds`() {
        // delete existing course if any
        try {
            courseService.deleteCourse("access-mock-course")
        } catch (_: ResponseStatusException) {
        }

        val path = "Mock-Course-Re"
        val file = File(path)
        val absolutePath = Paths.get(file.absolutePath)

        courseLifecycle.createFromDirectory(
            absolutePath,
            "https://github.com/mp-access/Mock-Course-Re.git"
        )

        courseService.getCourseBySlug("access-mock-course")
    }

    fun getCourse(): CourseWorkspace {
        return courseService.getCourseWorkspaceBySlug("access-mock-course")
    }

    @Test
    @WithMockUser(username="supervisor@uzh.ch", authorities = ["supervisor"])
    fun `Imported course slug correct`() {
        assertEquals("access-mock-course", getCourse().slug)
    }
    @Test
    @WithMockUser(username="supervisor@uzh.ch", authorities = ["supervisor"])
    fun `Imported course override end date correct`() {
        assertEquals(LocalDateTime.of(2028,1,1,13,0), getCourse().overrideEnd)
    }
    @Test
    @WithMockUser(username="supervisor@uzh.ch", authorities = ["supervisor"])
    fun `Imported course number of assignments correct`() {
        getCourse().assignments?.let { assertEquals(3, it.size) }
    }
    @Test
    @WithMockUser(username="supervisor@uzh.ch", authorities = ["supervisor"])
    fun `Imported course assignment numbers correct`() {
        assertEquals(setOf(1,2,3), getCourse().assignments?.map{ it?.ordinalNum }?.distinct()?.toSet())
    }

}

